### PR TITLE
chore: add property tests for encryption/decryption

### PR DESCRIPTION
**Pull request by olastor**
Hi, I created some basic property based tests using [fast-check](https://github.com/dubzzz/fast-check) in case that's helpful for this project. The tests run a bit longer than the regular unit tests since they do multiple iterations with random data. With the current setting of 50 runs per property it takes ca. 1000ms on my machine. I think the scrypt factor adds the most delay, so I could also change it to use a slower number for these tests. Happy to make more contributions if there's something to do and I find the time.

### CODE REVIEW

File: tests/property.test.ts [hunk 1]
```diff
@@ -0,0 +1,117 @@
...
+        const identity = generateIdentity()
+        const recipient = identityToRecipient(identity)
+
+        const enc = new Encrypter()
+        const dec = new Decrypter()
+
+        enc.addRecipient(recipient)
+        dec.addIdentity(identity)
+
+        return to_string(dec.decrypt(enc.encrypt(plaintext))) === plaintext
```
Comment (FiloSottile) [inline tests/property.test.ts:46]: Use `decrypt(..., "text")` instead of `to_string`.

File: tests/property.test.ts [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+    )
+  })
+
+
+  describe('Symmetric Encryption and Decryption', () => {
+    test.prop(
+      {
+        plaintext: fc.string(),
+        passphrase: fc.string(),
+        scryptWorkFactor: fc.integer({ min: 1, max: 12 })
```
Comment (FiloSottile) [inline tests/property.test.ts:79]: Let's cap the max a little lower, I don't think it's testing anything useful and will surely make things faster.

**Review by FiloSottile — COMMENTED** [review]
Sweet, thank you! A couple nits and then happy to merge.

More help is welcome, especially if there's something that the package is missing that I wouldn't know that would make it easier to use in more environments. (Like, does it bundle well for browsers? Should it have JSDocs?)

File: tests/property.test.ts (hunk 1 above)
Comment (olastor) [reply tests/property.test.ts]: Changed :heavy_check_mark:

File: tests/property.test.ts (hunk 2 above)
Comment (olastor) [reply tests/property.test.ts:78]: I lowered it to 4. If it's still to slow I could also lower the number of iterations.

**Comment by olastor** [comment]
> More help is welcome, especially if there's something that the package is missing that I wouldn't know that would make it easier to use in more environments. (Like, does it bundle well for browsers? Should it have JSDocs?)

@FiloSottile A few minor things I noticed when reading the code superficially:

- I think you probably want to add `"types": "./dist/index.d.ts"` to `package.json` to publish the type declarations (see [here](https://www.typescriptlang.org/docs/handbook/declaration-files/publishing.html#including-declarations-in-your-npm-package)).
- The code often seems to use `==` and `!=`, but it's best to avoid those in favor of the type-safe equality operators `===`/`!==` whenever possible. That's unintuitive when coming from other languages, but there are some [crazy edge cases](https://eslint.org/docs/latest/rules/eqeqeq) with the former operators.
- The way the classes/functions are exported seems a bit strange, but I guess that's mainly due sodium's ready function being called. I'll try to check if/how that could potentially be improved.

Regarding use in browser, I am not sure right now because I think sodium is using webasm. I think I could try to look into it, there are some packages like [argon2-browser](https://github.com/antelle/argon2-browser) which might be good to take as examples. Docs of the different function parameters would be good to have, I think (even if it's just part of the README).

**Comment by FiloSottile** [comment]
>     * I think you probably want to add `"types": "./dist/index.d.ts"` to `package.json` to publish the type declarations (see [here](https://www.typescriptlang.org/docs/handbook/declaration-files/publishing.html#including-declarations-in-your-npm-package)).

Sweet, thanks.

>     * The code often seems to use `==` and `!=`, but it's best to avoid those in favor of the type-safe equality operators `===`/`!==` whenever possible. That's unintuitive when coming from other languages, but there are some [crazy edge cases](https://eslint.org/docs/latest/rules/eqeqeq) with the former operators.

Done, thank you for pointing to the ESLint rule, I remembered this being a thing but I wanted to have it enforced with a tool.

>     * The way the classes/functions are exported seems a bit strange, but I guess that's mainly due sodium's ready function being called. I'll try to check if/how that could potentially be improved.

Yeah, this I went back and forth about a bunch. The issue is that libsodium is unusable (and unsafe) until the async sodium.ready is called. I wanted to make sure there was no way for a user to shoow themsleves in the foot by calling age functions before sodium.ready, and at the same time not make every age function async just because it might call sodium.ready. This is the best I could figure out, but I welcome suggestions, and it might need changing for browser bundling anyway.

> Regarding use in browser, I am not sure right now because I think sodium is using webasm. I think I could try to look into it, there are some packages like [argon2-browser](https://github.com/antelle/argon2-browser) which might be good to take as examples.

See #7.

**Pull request merged**